### PR TITLE
For Reverse P/Invokes, call TraceCall with the entrypoint to the target method.

### LINKED ISSUE #47376: Debugging BCL test projects won't load test assembly
I am a potential contributor to the .NET BCL. I've encountered an issue trying to debug BCL tests from the VS test runner.

### Description
I am unable to set breakpoints during debugging, unable to 'break all' (test will presumably continue running), unable to step through code as it will not break.

~~From the issues I've experienced, this problem shares similarities with issue #26732.~~

If a breakpoint is set beforehand, VS will inform me, hovering over it, that "No symbols have been loaded for this document", even though this is the test that is being debugged from the VS test runner. I assume this is the cause of my issues. All managed assemblies are of the Debug configuration.

Edit:
Attempting to find the tests assembly in the VS Modules window, I find it missing, along with the BCL library it should be testing.
The test(s) never complete no matter how much time its given.
Note: I'm debugging the tests for `System.Text.RegularExpressions`

### Configuration
System Config: Windows, x64
VS 2019 Version 4.8.04084

Trying to debug tests found in the master branch of this repository, after using the build command line tool to build everything.

### Other information
~~I've talked with people in the discord about this, people there suspect that its not loading the debug symbols because they aren't being put in a spot that VS can find it. May or may not be fixable from the build script.~~

I admit that this may or may not be a completely isolated issue. I'm still not able to solve it myself.

### LINKED ISSUE #47421: EEException on dotnet 6.0.100-preview.1.21074.5
### Description

Getting an System.ExecutionEngineException when running a WPF sample on dotnet 6.0.100-preview.1.21074.5, only when the debugger of VS 16.9.0 Preview 3.0 is attached.

This is the project.
https://github.com/microsoft/WPF-Samples/blob/master/Sample%20Applications/VideoViewerDemo/VideoViewerDemo.csproj
To make it run, I just made a 4 simple changes.

1) Update MainWindows.xaml, line 14, Directory, go back one more folder:
```xml
        <local:MyVideos Directory="../../../media" x:Key="Vids" />
```
2) Update csproj, targetframework to net6.0-windows, and change the project sdk to "Microsoft.NET.Sdk":
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net6.0-windows</TargetFramework>
...
```
3) Change the global.json file sdk version to match "6.0.100-preview.1.21074.5":
```json
{
  "sdk": {
    "version": "6.0.100-preview.1.21074.5",
```
4) Add nuget.config entry for dotnet6:
```xml
      <add key="dotnet6" value="https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet6/nuget/v3/index.json" />
```

### Configuration

* Which version of .NET is the code running on?
   * 6.0.100-preview.1.21074.5
* What is the architecture (x64, x86, ARM, ARM64)?
   * x64

### Other information

"An unhandled exception of type 'System.ExecutionEngineException' occurred in Unknown Module."

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Since we're in an IL stub, we don't need to support breakpoints in the stub itself. The debugger will try to resolve the target method and move the breakpoint there. In actuality, that's the part that's failing. Somewhere in the ReversePInvoke helper call chain, we're touching the secret arg register. As a result, when the debugger tries to move the breakpoint, it crashes. By doing the same work ourselves, the debugger doesn't need to resolve the stub secret arg since it never sees a breakpoint in the stub itself.

Fixes #47421 
Fixes #47376

### CODE REVIEW

**Review by hoyosjs — APPROVED** [review]

**Comment by jkotas** [comment]
> The debugger will try to resolve the target method and move the breakpoint there

Can the debugger be fixed to not resolve the target method in places where it is not supposed to instead?

**Comment by hoyosjs** [comment]
Looks like this didn't quite get the right helper called. The flag for IL stubs is not set early enough, so we still end up with said bad trace. The debugger is trying to resolve an int 3 location (my guess is it's VS's entrypoint, I still don't quite know why)

Stack:

``` 
1e 0000000f`d11bc2c0 00007ffa`80025bda     ntdll!KiUserExceptionDispatch+0x2e [minkernel\ntos\rtl\amd64\trampoln.asm @ 751] 
1f 0000000f`d11bca40 00007ffa`80026220     coreclr!UMEntryThunk::GetManagedTarget+0x6 [E:\repos\runtime3\src\coreclr\vm\dllimportcallback.h @ 213] 
20 0000000f`d11bca80 00007ffa`7fff9fb1     coreclr!ILStubManager::TraceManager+0xc0 [E:\repos\runtime3\src\coreclr\vm\stubmgr.cpp @ 1755] 
21 0000000f`d11bcab0 00007ffa`80128e16     coreclr!EEDbgInterfaceImpl::TraceManager+0x91 [E:\repos\runtime3\src\coreclr\vm\eedbginterfaceimpl.cpp @ 1261] 
22 0000000f`d11bcb60 00007ffa`8012677a     coreclr!DebuggerThreadStarter::TriggerPatch+0xa6 [E:\repos\runtime3\src\coreclr\debug\ee\controller.cpp @ 8248] 
23 0000000f`d11be9a0 00007ffa`80124b21     coreclr!DebuggerController::ScanForTriggers+0x1be [E:\repos\runtime3\src\coreclr\debug\ee\controller.cpp @ 2642] 
24 0000000f`d11bea20 00007ffa`801249ca     coreclr!DebuggerController::DispatchPatchOrSingleStep+0xd1 [E:\repos\runtime3\src\coreclr\debug\ee\controller.cpp @ 2937] 
25 0000000f`d11beb00 00007ffa`8012c566     coreclr!DebuggerController::DispatchNativeException+0x1ba [E:\repos\runtime3\src\coreclr\debug\ee\controller.cpp @ 4279] 
26 0000000f`d11beb50 00007ffa`7ff38874     coreclr!Debugger::FirstChanceNativeException+0xa6 [E:\repos\runtime3\src\coreclr\debug\ee\debugger.cpp @ 5627] 
27 0000000f`d11beba0 00007ffa`7ff11514     coreclr!IsDebuggerFault+0x64 [E:\repos\runtime3\src\coreclr\vm\excep.cpp @ 6575] 
28 0000000f`d11bebe0 00007ffa`7ff1146f     coreclr!CLRVectoredExceptionHandlerPhase2+0x94 [E:\repos\runtime3\src\coreclr\vm\excep.cpp @ 7282] 
29 0000000f`d11bec20 00007ffa`7ff113aa     coreclr!CLRVectoredExceptionHandler+0x8b [E:\repos\runtime3\src\coreclr\vm\excep.cpp @ 7198] 
2a 0000000f`d11bec80 00007ffb`03aa83dc     coreclr!CLRVectoredExceptionHandlerShim+0xea [E:\repos\runtime3\src\coreclr\vm\excep.cpp @ 7937] 
2b 0000000f`d11becd0 00007ffb`03a7b406     ntdll!RtlpCallVectoredHandlers+0x108 [minkernel\ntdll\vectxcpt.c @ 204] 
2c (Inline Function) --------`--------     ntdll!RtlCallVectoredExceptionHandlers+0xe [minkernel\ntdll\vectxcpt.c @ 345] 
2d 0000000f`d11bed70 00007ffb`03acfe3e     ntdll!RtlDispatchException+0x66 [minkernel\ntos\rtl\amd64\exdsptch.c @ 357] 
2e 0000000f`d11bef80 00007ffa`203cc75a     ntdll!KiUserExceptionDispatch+0x2e [minkernel\ntos\rtl\amd64\trampoln.asm @ 751] 
2f 0000000f`d11bf720 00007ffb`03a3ad4f     System_Private_CoreLib!ILStubClass.IL_STUB_ReversePInvoke(IntPtr, Int32, Byte, Int64, Int64, Int64, Int64)+0x2a
30 0000000f`d11bf7b0 00007ffb`03a3a96d     ntdll!EtwpEventApiCallback+0xdb [minkernel\etw\ntdll\evntapi.c @ 283] 
```

**Comment by jkotas** [comment]
Fix `ILStubManager` to kick in only for IP at the start of the method?

**Comment by hoyosjs** [comment]
It seems like it was the ILStubManager itself trying to trace ILStubClass.IL_STUB_ReversePInvoke, it emitted a TRACE_MGR_PUSH and we put the int 3 at that point of the streamto try to continue tracing.

File: src/coreclr/vm/jithelpers.cpp [hunk 1]
```diff
@@ -5445,12 +5445,16 @@ void F_CALL_CONV HCCALL3(JIT_ReversePInvokeEnterTrackTransitions, ReversePInvoke
         thread->m_fPreemptiveGCDisabled.StoreWithoutBarrier(1);
         if (g_TrapReturningThreads.LoadWithoutBarrier() != 0)
         {
-            JIT_ReversePInvokeEnterRare2(frame, _ReturnAddress());
+            // If we're in an IL stub, we want to trace the address of the target method,
+            // not the next instruction in the stub.
+            JIT_ReversePInvokeEnterRare2(frame, pMD->IsILStub() && secretArg != NULL ? (void*)frame->pMD->GetMultiCallableAddrOfCode() : _ReturnAddress());
```
Comment (jkotas) [inline src/coreclr/vm/jithelpers.cpp:5450]: It is dangerous to call `MethodDesc::GetMultiCallableAddrOfCode`. This method can take locks and throw exceptions.

Comment (jkoritzinsky) [reply src/coreclr/vm/jithelpers.cpp]: I've updated this to use the same code-path that the old system used.

**Comment by jkotas** [comment]
> It seems like it was the ILStubManager itself trying to trace

Right. It tells me that it may be better to fix `ILStubManager` to not get confused. Is there a good reason why we are trying to fix this in the callouts and not in the `ILStubManager`?

**Review by hoyosjs — APPROVED** [review]

**Comment by hoyosjs** [comment]
Merging this to unblock debugging in the rest of the stack.

**Pull request merged**